R/code_feedback.R

```diff
@@ -218,6 +235,47 @@ with_maybe_code_feedback <- function(val, expr) {
   )
 }
 
+which_closest_solution_code <- function(
+  user_code = .user_code,
+  solution_code_all = .solution_code_all
+) {
+  if (is_placeholder(solution_code_all, ".solution_code_all")) {
+    solution_code_all <- get0(".solution_code_all", parent.frame())
+    assert_not_placeholder(solution_code_all)
+  }
+
+  # If there's no solution code or only one solution,
+  # we don't need to find the closest match
+  if (length(solution_code_all) < 2) {
+    return(length(solution_code_all))
+  }
+
+  if (is_placeholder(user_code, ".user_code")) {
+    user_code <- get0(".user_code", parent.frame())
+    assert_not_placeholder(user_code)
+  }
+
+  # Convert from list to character vector
+  solution_code_all <- unlist(solution_code_all)
+
+  standardise_code_text <- function(code) {
+    code %>%
+      unpipe_all_str() %>%
+      rlang::parse_exprs() %>%
+      call_standardise_formals_recursive() %>%
+      purrr::map_chr(rlang::expr_text) %>%
+      paste(collapse = "\n")
+  }
+
+  user_code <- standardise_code_text(user_code)
+
+  solution_code_all <- solution_code_all %>%
+    purrr::map_chr(standardise_code_text)
+
+  # Find the index of the solution code that the user code is closest to
+  which.min(utils::adist(user_code, solution_code_all))
```

Comment:
If there is a tie, `which.min()` returns the first index in the tie, e.g.
``` r
options <- c("abc", "y", "z")
which.min(adist("x", options))
#> [1] 2
```

I don't know about returning `integer(0)` if there's a tie, because we would lose information if e.g. there were five options and the tie was only between two of them.

Since `.solution` defaults to the last value in `.solution_code_all()`, we could have it return the last tie index instead of the first tie index. That way if `.solution` is one of the tied values, that would always be the one returned. If `.solution` isn't one of the tied values, it seems reasonable to choose one arbitrarily.

Reply:
Sorry, I realized later ([last comment](https://github.com/rstudio/gradethis/pull/289#discussion_r818142324)) that there are two tie scenarios. We'd only return `integer(0)` if the tie is between _all options_.

Otherwise, we'd return the first in the tie — unless the tie includes `.solution` (i.e. the last value).

So yeah, this would all require replacing `which.min()` with our own logic.

Reply:
Is it okay to always return the _last_ value if there's a tie? If we can do that, we could simplify the logic, because `.solution` is always the last value.

Reply:
> Is it okay to always return the _last_ value if there's a tie?

No, just when the last value is included in the tie

Reply:
Implemented in [658ef2c](https://github.com/rstudio/gradethis/pull/289/commits/658ef2c58afd757863580ea0cc937e8c8f0d4371), but we can discuss further if needed

Reply:
@gadenbuie To clarify, if we have a vector with length 5 and the tie is between 2 and 4, is it necessary to return 2, or is it also acceptable to return 4? If it is okay to return 4, then it would simplify the logic, because a tie including between 2 and 5 (the last value) wouldn't need to be treated any different than a tie between 2 and 4.

Reply:
Yeah I thought we would return 2. It's a little convoluted, but I think ranking solutions top-bottom (top first) makes sense — other than the special case of the "default" solution.

I'd rather make the first solution the default than generally go with a bottom-top preference order.